--- a/feeds/luci/modules/luci-base/root/usr/libexec/rpcd/luci
+++ b/feeds/luci/modules/luci-base/root/usr/libexec/rpcd/luci
@@ -46,7 +46,10 @@
 	getCPUInfo = {
 		call = function()
 			local sys = require "luci.sys"
-			return { result = sys.exec("top -n1 | awk '/^CPU/ {printf(\"%d%\",100-$8)}'") }
+			local rv = {}
+			rv.cpufreq = sys.exec("echo $(expr $(cat /sys/devices/system/cpu/cpufreq/policy0/cpuinfo_cur_freq) / 1000) MHz +$(awk '{printf(\"%.1f°C \",$0/1000)}' /sys/class/ieee80211/phy*/device/hwmon/hwmon*/temp1_input)")
+			rv.cpufree = sys.exec("top -n1 | awk '/^CPU/ {printf(\"%d%\",100-$8)}'")
+			return rv
 		end
 	},
 

--- a/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/10_system.js
+++ b/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/10_system.js
@@ -62,6 +62,7 @@
 			_('Model'),            boardinfo.model,
 			_('Architecture'),     boardinfo.system,
 			_('Target Platform'),  (L.isObject(boardinfo.release) ? boardinfo.release.target : ''),
+			_('CPU Info'),         cpuinfo.cpufreq,
 			_('Firmware Version'), (L.isObject(boardinfo.release) ? boardinfo.release.description + ' / ' : '') + (luciversion || ''),
 			_('Kernel Version'),   boardinfo.kernel,
 			_('Local Time'),       datestr,
@@ -71,7 +72,7 @@
 				systeminfo.load[1] / 65535.0,
 				systeminfo.load[2] / 65535.0
 			) : null,
-			_('CPU usage (%)'),    cpuinfo.result
+			_('CPU usage (%)'),    cpuinfo.cpufree
 		];
 
 		var table = E('table', { 'class': 'table' });

--- a/package/ipk/luci-app-cpufreq/root/etc/config/cpufreq
+++ b/package/ipk/luci-app-cpufreq/root/etc/config/cpufreq
@@ -1,2 +1,6 @@
 
 config settings 'cpufreq'
+	option mode0 'schedutil'
+	option min0 '48000'
+	option max0 '896000'
+

--- a/target/linux/ipq40xx/base-files/etc/board.d/01_leds
+++ b/target/linux/ipq40xx/base-files/etc/board.d/01_leds
@@ -13,13 +13,14 @@
 	ucidef_set_led_netdev "wan" "WAN" "amber:wan" "eth1"
 	;;
 asus,rt-ac42u)
+	ucidef_set_led_switch "wan" "WAN(blue)" "blue:wan" "switch0" "0x20"
 	ucidef_set_led_switch "lan1" "LAN1" "blue:lan-1" "switch0" "0x02"
 	ucidef_set_led_switch "lan2" "LAN2" "blue:lan-2" "switch0" "0x04"
 	ucidef_set_led_switch "lan3" "LAN3" "blue:lan-3" "switch0" "0x08"
 	ucidef_set_led_switch "lan4" "LAN4" "blue:lan-4" "switch0" "0x10"
 	;;
 asus,rt-ac58u)
-	ucidef_set_led_netdev "wan" "WAN" "blue:wan" "eth1"
+	ucidef_set_led_switch "wan" "WAN" "blue:wan" "switch0" "0x20"
 	ucidef_set_led_switch "lan" "LAN" "blue:lan" "switch0" "0x1e"
 	;;
 avm,fritzbox-4040)

--- a/target/linux/ipq40xx/base-files/etc/board.d/02_network
+++ b/target/linux/ipq40xx/base-files/etc/board.d/02_network
@@ -58,13 +58,9 @@
 	asus,rt-ac42u|\
 	asus,rt-ac58u|\
 	mikrotik,hap-ac2|\
-	mikrotik,hap-ac3|\
-	p2w,r619ac-64m|\
-	p2w,r619ac-128m|\
-	zyxel,nbg6617)
-		ucidef_set_interfaces_lan_wan "eth0" "eth1"
+	mikrotik,hap-ac3)
 		ucidef_add_switch "switch0" \
-			"0u@eth0" "1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1"
+			"0t@eth0" "1:lan" "2:lan" "3:lan" "4:lan" "5:wan"
 		;;
 	avm,fritzbox-4040|\
 	linksys,ea6350v3|\
@@ -151,6 +147,12 @@
 		ucidef_add_switch "switch0" \
 			"0u@eth0" "2:lan:4" "3:lan:3" "4:lan:2" "0u@eth1" "5:wan"
 		;;
+	p2w,r619ac-64m|\
+	p2w,r619ac-128m|\
+	zyxel,nbg6617)
+		ucidef_add_switch "switch0" \
+			"0t@eth0" "1:lan:4" "2:lan:3" "3:lan:2" "4:lan:1" "5:wan:5"
+		;;
 	*)
 		echo "Unsupported hardware. Network interfaces not initialized"
 		;;
@@ -168,10 +170,11 @@
 	8dev,habanero-dvk)
 		label_mac=$(mtd_get_mac_binary "ART" 0x1006)
 		;;
+	asus,rt-ac42u|\
 	asus,rt-ac58u)
-		wan_mac=$(mtd_get_mac_binary_ubi Factory 0x1006)
-		lan_mac=$(mtd_get_mac_binary_ubi Factory 0x5006)
-		label_mac=$wan_mac
+		lan_mac=$(cat /sys/class/net/eth0/address)
+		wan_mac=$(macaddr_add "$lan_mac" 1)
+		label_mac=$lan_mac
 		;;
 	cilab,meshpoint-one)
 		label_mac=$(mtd_get_mac_binary "ART" 0x1006)
@@ -217,6 +220,10 @@
 		lan_mac=$(cat /sys/firmware/mikrotik/hard_config/mac_base)
 		label_mac="$lan_mac"
 		;;
+	p2w,r619ac-64m|\
+	p2w,r619ac-128m)
+		wan_mac=$(cat /sys/class/net/eth0/address)
+		lan_mac=$(macaddr_add "$wan_mac" 1)
 	esac
 
 	[ -n "$lan_mac" ] && ucidef_set_interface_macaddr "lan" $lan_mac

--- a/target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh
+++ b/target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh
@@ -7,9 +7,10 @@
 		ip link set dev eth0 address $(macaddr_add "$base_mac" 1)
 		ip link set dev eth1 address $(macaddr_add "$base_mac" 3)
 		;;
-	asus,rt-ac42u)
-		ip link set dev eth0 address $(mtd_get_mac_binary_ubi Factory 0x1006)
-		ip link set dev eth1 address $(mtd_get_mac_binary_ubi Factory 0x9006)
+	asus,rt-ac42u|\
+	asus,rt-ac58u)
+		base_mac=$(mtd_get_mac_binary_ubi Factory 0x1006)
+		ip link set dev eth0 address $(macaddr_add "$base_mac" 1)
 		;;
 	engenius,eap2200)
 		base_mac=$(cat /sys/class/net/eth0/address)

--- a/target/linux/ipq40xx/files/drivers/net/ethernet/qualcomm/essedma/edma.c
+++ b/target/linux/ipq40xx/files/drivers/net/ethernet/qualcomm/essedma/edma.c
@@ -727,13 +727,11 @@
 			edma_receive_checksum(rd, skb);
 
 			/* Process VLAN HW acceleration indication provided by HW */
-			if (unlikely(adapter->default_vlan_tag != rd->rrd4)) {
-				vlan = rd->rrd4;
-				if (likely(rd->rrd7 & EDMA_RRD_CVLAN))
-					__vlan_hwaccel_put_tag(skb, htons(ETH_P_8021Q), vlan);
-				else if (rd->rrd1 & EDMA_RRD_SVLAN)
-					__vlan_hwaccel_put_tag(skb, htons(ETH_P_8021AD), vlan);
-			}
+			vlan = rd->rrd4;
+			if (likely(rd->rrd7 & EDMA_RRD_CVLAN))
+				__vlan_hwaccel_put_tag(skb, htons(ETH_P_8021Q), vlan);
+			else if (rd->rrd1 & EDMA_RRD_SVLAN)
+				__vlan_hwaccel_put_tag(skb, htons(ETH_P_8021AD), vlan);
 
 			/* Update rx statistics */
 			adapter->stats.rx_packets++;
