--- a/src/dns_conf.c
+++ b/src/dns_conf.c
@@ -603,10 +603,9 @@ static int _config_rule_group_setup_value(struct dns_conf_group_info *group_info
 
 	memset(soa_table, 0, soa_talbe_size);
 	memcpy(&group_rule->check_orders, &dns_conf_default_check_orders, sizeof(group_rule->check_orders));
-	group_rule->dualstack_ip_selection = 1;
+	group_rule->dualstack_ip_selection = 0;
 	group_rule->dns_dualstack_ip_selection_threshold = 10;
-	group_rule->dns_rr_ttl_min = 600;
-	group_rule->dns_serve_expired = 1;
+	group_rule->dns_serve_expired = 0;
 	group_rule->dns_serve_expired_ttl = 24 * 3600 * 3;
 	group_rule->dns_serve_expired_reply_ttl = 3;
 	group_rule->dns_max_reply_ip_num = DNS_MAX_REPLY_IP_NUM;
