--- a/feeds/luci/modules/luci-base/root/usr/share/rpcd/ucode/luci
+++ b/feeds/luci/modules/luci-base/root/usr/share/rpcd/ucode/luci
@@ -74,7 +74,10 @@
 
 	getCPUInfo: {
 		call: function() {
-			return { result: popen('top -n1 | awk \'/^CPU/ {printf("%d%",100-$8)}\'').read('all') };
+			let rv = {};
+			rv.cpufreq = popen('awk \'{printf("%.1f°C",$0/1000)}\' /sys/class/thermal/thermal_zone0/temp').read('all');
+			rv.cpufree = popen('top -n1 | awk \'/^CPU/ {printf("%d%",100-$8)}\'').read('all');
+			return rv;
 		}
 	},
 

--- a/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/10_system.js
+++ b/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/10_system.js
@@ -63,6 +63,7 @@
 			_('Model'),            boardinfo.model,
 			_('Architecture'),     boardinfo.system,
 			_('Target Platform'),  (L.isObject(boardinfo.release) ? boardinfo.release.target : ''),
+			_('CPU Info'),         cpuinfo.cpufreq,
 			_('Firmware Version'), (L.isObject(boardinfo.release) ? boardinfo.release.description + ' / ' : '') + (luciversion || ''),
 			_('Kernel Version'),   boardinfo.kernel,
 			_('Local Time'),       datestr,
@@ -72,7 +73,7 @@
 				systeminfo.load[1] / 65535.0,
 				systeminfo.load[2] / 65535.0
 			) : null,
-			_('CPU usage (%)'),    cpuinfo.result
+			_('CPU usage (%)'),    cpuinfo.cpufree
 		];
 
 		var table = E('table', { 'class': 'table' });

--- a/package/ipk/luci-app-flowoffload/root/etc/config/flowoffload
+++ b/package/ipk/luci-app-flowoffload/root/etc/config/flowoffload
@@ -1,4 +1,5 @@
 
 config flowoffload
 	option enabled '1'
+	option hw '1'
 

--- a/target/linux/mediatek/Makefile
+++ b/target/linux/mediatek/Makefile
@@ -8,8 +8,7 @@
 SUBTARGETS:=mt7622 mt7623 mt7629 filogic
 FEATURES:=dt-overlay emmc fpu gpio nand pci pcie rootfs-part separate_ramdisk squashfs usb
 
-KERNEL_PATCHVER:=5.15
-KERNEL_TESTING_PATCHVER:=6.1
+KERNEL_PATCHVER:=6.1
 
 include $(INCLUDE_DIR)/target.mk
 DEFAULT_PACKAGES += \
