--- a/r8152.c
+++ b/r8152.c
@@ -1027,6 +1027,15 @@
 #define RTL_ADVERTISED_1000_FULL		BIT(5)
 #define RTL_ADVERTISED_2500_FULL		BIT(6)
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 18, 0)
+static inline void netif_set_gso_max_size(struct net_device *dev,
+					  unsigned int size)
+{
+	/* dev->gso_max_size is read locklessly from sk_setup_caps() */
+	WRITE_ONCE(dev->gso_max_size, size);
+}
+#endif
+
 /* Maximum number of multicast addresses to filter (vs. Rx-all-multicast).
  * The RTL chips use a 64 element hash table based on the Ethernet CRC.
  */
@@ -20723,10 +20732,17 @@
 
 	usb_set_intfdata(intf, tp);
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 1, 0)
 	if (tp->support_2500full)
 		netif_napi_add(netdev, &tp->napi, r8152_poll, 256);
 	else
 		netif_napi_add(netdev, &tp->napi, r8152_poll, 64);
+#else
+	if (tp->support_2500full)
+		netif_napi_add_weight(netdev, &tp->napi, r8152_poll, 256);
+	else
+		netif_napi_add_weight(netdev, &tp->napi, r8152_poll, 64);
+#endif
 
 	ret = register_netdev(netdev);
 	if (ret != 0) {
